libraries/ms-common/src/stm32f0xx/interrupt.c

```diff
@@ -0,0 +1,200 @@
+  }
+
+  EXTI_ClearITPendingBit(0x01 << line);
+  return STATUS_CODE_OK;
+}
+
+// GPIO INTERRUPTS
+
+static interrupt_callback s_interrupt_gpio_callback[NUM_GPIO_PINS];
+
+// Helper function to enable gpio interrupts.
+StatusCode prv_gpio_interrupt_enable(GPIOAddress *address, InterruptSettings *settings,
+                                     interrupt_callback callback, uint8_t irq_channel) {
+  SYSCFG_EXTILineConfig(address->port, address->pin);
+  status_ok_or_return(prv_exti_enable(address->pin, settings));
+  s_interrupt_gpio_callback[address->pin] = callback;
+  status_ok_or_return(prv_nvic_enable(irq_channel, settings->priority));
+  return STATUS_CODE_OK;
+}
+
+StatusCode interrupt_gpio_register(GPIOAddress *address, InterruptSettings *settings,
+                                   interrupt_callback callback) {
+  if (address->port >= NUM_GPIO_PORTS || address->pin >= NUM_GPIO_PINS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  } else if (s_interrupt_gpio_callback[address->pin] != NULL) {
+    return status_msg(STATUS_CODE_RESOURCE_EXHAUSTED, "Interrupt for this pin is already used.");
+  }
+
+  if (address->pin < 2) {
+    // Pins 0, 1 are on NVIC channel 5 which is gpio priority group 0.
+    if (s_interrupt_gpio_priority_map[0] != settings->priority &&
+        s_interrupt_gpio_priority_map[0] != NUM_INTERRUPT_PRIORITY) {
+      return status_msg(STATUS_CODE_INVALID_ARGS, "Priority already set for pins 0-1.");
+    }
+    status_ok_or_return(prv_gpio_interrupt_enable(address, settings, callback, 5));
+  } else if (address->pin < 4) {
+    // Pins 2, 3 are on NVIC channel 6 which is gpio priority group 1.
+    if (s_interrupt_gpio_priority_map[1] != settings->priority &&
+        s_interrupt_gpio_priority_map[1] != NUM_INTERRUPT_PRIORITY) {
+      return status_msg(STATUS_CODE_INVALID_ARGS, "Priority already set for pins 2-3.");
+    }
+    status_ok_or_return(prv_gpio_interrupt_enable(address, settings, callback, 6));
+  } else {
+    // Pins 4, 15 are on NVIC channel 7 which is gpio priority group 2.
+    if (s_interrupt_gpio_priority_map[2] != settings->priority &&
+        s_interrupt_gpio_priority_map[2] != NUM_INTERRUPT_PRIORITY) {
+      return status_msg(STATUS_CODE_INVALID_ARGS, "Priority already set for pins 4-15.");
+    }
+    status_ok_or_return(prv_gpio_interrupt_enable(address, settings, callback, 7));
+  }
+  return STATUS_CODE_OK;
+}
+
+StatusCode interrupt_gpio_trigger(GPIOAddress *address) {
+  if (address->port >= NUM_GPIO_PORTS || address->pin >= NUM_GPIO_PINS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  prv_exti_trigger_interrupt(address->pin);
+  return STATUS_CODE_OK;
+}
+
+// Callback runner for GPIO which runs callbacks based on which callbacks are associated with an IRQ
+// channel. The function runs the callbacks which have a flag raise in the range [lower_bound,
+// upperbound).
+void prv_run_gpio_callbacks(uint8_t lower_bound, uint8_t upper_bound) {
```

Comment:
That's how most for loops using < and range functions I've even seen are written but I can change it if it matters that much.